js/script.js

```diff
@@ -0,0 +1,199 @@
+    .map(el => `${el.title} c рейтингом ${el.imdbRating}`)
+
+// 4. Создать новый массив, где объекты фильмов будут состоять из следующих полей: 
+//     `id, title, released, plot`
+
+filmInfo = mainArray.reduce((total, {id, title, released, plot}) => {
+    return [...total, { id, title, released, plot }]
+}, []);
+
+// 5. Создать объект, где ключ это имя актера, а значение - массив из фильмов с его участием
+
+// id: 3,
+//         title: "Star Wars",
+//         year: 1977,
+//         released: "25 May 1977",
+//         actors: ["Mark Hamill", "Harrison Ford", "Carrie Fisher"],
+//         plot: "Luke Skywalker joins forces with a Jedi Knight, a cocky pilot, a Wookiee and two droids to save the galaxy from the Empire's world-destroying battle station, while also attempting to rescue Princess Leia from the mysterious Darth Vad",
+   
+let actorsInfo = mainArray
+    .reduce((total, {title, actors}) => {
+    return [...total, {title, actors}]
+    }, [])
+    .reduce((total, currObj, i, arr)=>{
+        let actorsArr = arr.map(el => el.actors).flat(Infinity).filter((el, i, arr) => (arr.indexOf(el) === i));
+        for(let val of actorsArr){
+            if(currObj.actors.includes(val)){
+                if (total.hasOwnProperty(val)) {
+                    return {
+                      ...total,
+                      [val]: [...total[val], currObj.title],
+                    };
+                  }  
+                   return {
+                    ...total, [val]: [currObj.title]
+                   }
+            }  
+        }
+    }, {}) 
+;
+
+
+    console.log(actorsInfo)
+// 6. Создать массив авторов (поле writer) без повторений.
+
+writerArray = mainArray
+     .map(el => el.writer).flat(Infinity).filter((el, i, arr) => (arr.indexOf(el) === i));
+
+// 7. Создать функцию, которая бы принимала массив фильмов и строку. А результатом этой функции должен быть новый отфильтрованный массив, с фильмами, где строка входит в название фильма.
+    
+    function getFilmsTitles(filmArray, currStr){
+        return filmArray.filter(({title}) => title.indexOf(currStr) > -1)
+    
+    }
+    
+
+// 8. Создать функцию, которая бы принимала массив фильмов и число. А результатом этой функции должен быть отфильтрованный массив, с фильмами где число равно году выхода фильма.
+    
+    function getFilmsYear(filmArray, currYear){
+        return filmArray.reduce((total, {title, year}) => {
+            if(year === currYear){
+                return [ ...total,{ [title]: year } ]
+            }
+         return total;
+        }, [])
+    }
+    
+// 9. Создать функцию, которая бы принимала массив фильмов и строку. А результатом этой функции должен быть отфильтрованный массив, с фильмами где строка входит в название фильма или в его сюжет.
+
+    function checkFilmsTitlesOrPlot(filmArray, currStr){
+        return filmArray.filter(({title, plot}) => ((title.indexOf(currStr) > -1) || (plot.indexOf(currStr) > -1)))
```

Comment:
the same problem
'Black Widow' -> positive result,
but 'black widow' -> negative